Add a new Batched NMS OP

supriyar opened the pull request:
This change adds a new Batched NMS operation (called NonMaxSuppressionLite). The entire NMS operation for all batches and across all classes is performed in one kernel operation.
The change also adds unit tests for the new kernel operation.

Commit:
First Implementation of basic batched non_max_suppression kernel and unit tests.
Works for all unit tests with batch size 1 and 2.

Commit:
Add NMSLite shape function

Commit:
Added support for padding to the max_total_size (per batch). Output is padded
to zeros. The pad_to_max_total_size is an optional bool argument. The
valid_detections output must be checked to get the number of valid outputs per
batch.

Commit:
Updated with Debug messages (for dev effort only) and added clip box param

Commit:
Remove debug prints

Commit:
Updated kernel unit tests to account for clipped bounding box

Commit:
Add support for use_static_shapes.
If false, the output is always padded to max_total_size
If true, the output is padded to the minimum of max_total_size and
max_size_per_class * num_classes
Updated unit tests to take this into account

Commit:
Fix a bug in calculating max_detections

Commit:
Added argument description, removed trailing whitespaces

supriyar commented:
I signed it!

azaks2 commented on tensorflow/core/kernels/non_max_suppression_op.cc:
```diff
@@ -195,6 +222,223 @@ void DoNonMaxSuppressionOp(
 }
 
+void BatchedNonMaxSuppressionOp(OpKernelContext* context, 
+                           const Tensor& inp_boxes, const Tensor& inp_scores, 
```
align params and do 1 per line

supriyar commented on tensorflow/core/ops/image_ops.cc:
```diff
@@ -131,6 +131,53 @@ Status NMSShapeFn(InferenceContext* c) {
+  TF_RETURN_IF_ERROR(c->WithRank(c->input(1), 3, &scores));
+  ShapeHandle max_output_size_per_class;
+  TF_RETURN_IF_ERROR(c->WithRank(c->input(2), 0, &max_output_size_per_class));
+  ShapeHandle max_total_size;
```
max_total_size is the total output size per batch. If use_static_shapes is false, then we always pad to max_total_size.

supriyar replied:
This statement just checks the rank of the max_total_size input. I'm doing similar checks for other inputs as well. Let me know if that's not the correct way.

Commit:
Align function parameters, remove trailing whitespace

supriyar commented on tensorflow/python/ops/image_ops_impl.py:
```diff
@@ -2808,3 +2808,68 @@ def sobel_edges(image):
+      overlap too much with respect to IOU.
+    score_threshold: A float representing the threshold for deciding when to
+      remove boxes based on score.
+    use_static_shapes: If true, the output nmsed boxes, scores, classes and 
```
Based on the current implementation, if use_static_shapes is false, we always pad/clip to max_total_size.
https://github.com/tensorflow/models/blob/628b970a3d7c59a3b65220e24972f9987e879bca/research/object_detection/core/post_processing.py#L454
And if use_static_shapes is true we pad to minimum of (max_total_size, max_size_per_class*num_classes). Please let me know if our understanding is incorrect.

Also, looks like max_total_size can never be 0 or less than max_size_per_class 
https://github.com/tensorflow/models/blob/c4c49d1a1534b3664ac12aad69a124f87a9e3f7f/research/object_detection/builders/post_processing_builder.py#L78

azaks2 replied:
your understanding is most likely correct. But the op is more generic that the python function.
As long as we can reimplement the python using the new op we are good to go. Having use_static_shapes == pad or truncate seems less confusing. A client can always pass in max_total_size = max_size_per_class*num_classes if desired

supriyar replied:
The goal here is to implement the python function `batch_multiclass_non_max_suppression` (in post_processing.py) using the new op. So in that case I'm keeping the explanation as is, unless there is something I'm missing.

azaks2 replied:
so may be rename use_static_shapes to pad_per_class

supriyar commented on tensorflow/core/ops/image_ops.cc:
```diff
@@ -131,6 +131,53 @@ Status NMSShapeFn(InferenceContext* c) {
+    DimensionHandle output_dim;
+    DimensionHandle batch_dim = c->UnknownDim();
+
+    TF_RETURN_IF_ERROR(c->MakeDimForScalarInput(3, &output_dim));
```
Added a check for max_total_size to be > 0

supriyar commented on tensorflow/core/ops/image_ops.cc:
```diff
@@ -131,6 +131,53 @@ Status NMSShapeFn(InferenceContext* c) {
+  TF_RETURN_IF_ERROR(c->GetAttr("use_static_shapes", &use_static_shapes));
+  if(!use_static_shapes) {
+    DimensionHandle output_dim;
+    DimensionHandle batch_dim = c->UnknownDim();
```
Updated

Commit:
Addressed review comments. Made the following main changes
1. Added calculation of batch_dim in NMSLiteShapeFn
2. Added a check for max_total_size to be > 0
3. Updated clip calculation to account for all clip co-ordinates
4. Other code style related changes

Commit:
Updated name of use_static_shapes to pad_per_class to better reflect the
operation

azaks2 commented on tensorflow/core/ops/image_ops.cc:
```diff
@@ -778,4 +825,19 @@ REGISTER_OP("NonMaxSuppressionWithOverlaps")
       return Status::OK();
     });
 
+REGISTER_OP("NonMaxSuppressionLite")
```
CombinedNonMaxSupression

azaks2 commented on tensorflow/core/ops/image_ops.cc:
```diff
@@ -131,6 +131,56 @@ Status NMSShapeFn(InferenceContext* c) {
+  TF_RETURN_IF_ERROR(c->WithRank(c->input(5), 0, &score_threshold));
+
+  DimensionHandle unused;
+  //boxes[1] and scores[1] are both num_anchors
```
check batch_size and q dimensions as well

azaks2 commented on tensorflow/core/ops/image_ops.cc:
```diff
@@ -131,6 +131,56 @@ Status NMSShapeFn(InferenceContext* c) {
+
+  bool pad_per_class;
+  TF_RETURN_IF_ERROR(c->GetAttr("pad_per_class", &pad_per_class));
+  if(!pad_per_class) {
```
space after if

azaks2 commented on tensorflow/core/ops/image_ops.cc:
```diff
@@ -131,6 +131,56 @@ Status NMSShapeFn(InferenceContext* c) {
+    DimensionHandle batch_dim = c->Dim(boxes, 0);
+    
+    TF_RETURN_IF_ERROR(c->MakeDimForScalarInput(3, &output_dim));
+    if(c->ValueKnown(output_dim) && c->Value(output_dim) <= 0) {
```
space after if

azaks2 commented on tensorflow/core/ops/image_ops.cc:
```diff
@@ -131,6 +131,56 @@ Status NMSShapeFn(InferenceContext* c) {
+    DimensionHandle output_dim;
+    DimensionHandle batch_dim = c->Dim(boxes, 0);
+    
+    TF_RETURN_IF_ERROR(c->MakeDimForScalarInput(3, &output_dim));
```
can you use ShapeHandle max_total_size

azaks2 commented on tensorflow/core/ops/image_ops.cc:
```diff
@@ -131,6 +131,56 @@ Status NMSShapeFn(InferenceContext* c) {
+  else {
+    //TODO SR Can all the outputs be unknowndim?
+    c->set_output(0, c->Vector(c->UnknownDim()));
+    c->set_output(1, c->Vector(c->UnknownDim()));
```
we do know the output size:

min(max_output_size, max_output_size_per_class*classes)

Commit:
Added more checks for NMS function. Updated name of the new NMS op to
CombinedNonMaxSuppression

azaks2 commented on tensorflow/core/ops/image_ops.cc:
```diff
@@ -131,6 +131,71 @@ Status NMSShapeFn(InferenceContext* c) {
+
+  DimensionHandle d = c->Dim(boxes, 2);
+  auto num_classes = c->Value(c->Dim(scores, 2));
+  if (c->ValueKnown(d)) {
```
why guard with ValueKnown here but not at line 159?

azaks2 commented on tensorflow/core/ops/image_ops.cc:
```diff
@@ -131,6 +131,71 @@ Status NMSShapeFn(InferenceContext* c) {
+
+  TF_RETURN_IF_ERROR(c->MakeDimForScalarInput(3, &output_dim));
+  if (c->ValueKnown(output_dim) && c->Value(output_dim) <= 0) {
+    return errors::InvalidArgument("max_total_size should be > 0 ");
```
should we just default to size_per_class * num_classes. a negative value would mean a user does not intended to cap.
we can check that par_per_class is true.

supriyar replied:
https://github.com/tensorflow/models/blob/master/research/object_detection/builders/post_processing_builder.py#L78
Fails if max_total_size is less than size_per_class. Even if pad_per_class is true, we still need max_total_size as it might be smaller than size_per_class * num_classes. 
If the user does not intend to cap, they can set pad_per_class to true and max_total_size to > size_per_class * num_classes. Does that sound reasonable?

azaks2 commented on tensorflow/python/ops/image_ops_impl.py:
```diff
@@ -3058,3 +3058,69 @@ def crop_and_resize_v2(
+
+@tf_export('image.combined_non_max_suppression')
+def combined_non_max_suppression(boxes,
+                             scores,
```
align below boxes

azaks2 commented on tensorflow/python/ops/image_ops_impl.py:
```diff
@@ -3058,3 +3058,69 @@ def crop_and_resize_v2(
+  [y1, x1, y2, x2], where (y1, x1) and (y2, x2) are the coordinates of any
+  diagonal pair of box corners and the coordinates can be provided as normalized
+  (i.e., lying in the interval [0, 1]) or absolute.  Note that this algorithm
+  is agnostic to where the origin is in the coordinate system.  Note that this
```
s/Note/Also note

azaks2 commented on tensorflow/core/kernels/non_max_suppression_op.cc:
```diff
@@ -195,6 +222,229 @@ void DoNonMaxSuppressionOp(
+
+  int q = inp_boxes.dim_size(2);
+  int num_classes = inp_scores.dim_size(2);
+  //unpack along batch dimension
```
space after //

azaks2 replied:
also we do not really unpack anything here

azaks2 commented on tensorflow/core/kernels/non_max_suppression_op.cc:
```diff
@@ -195,6 +222,229 @@ void DoNonMaxSuppressionOp(
+  // Default clip window of [0, 0, 1, 1] if none specified
+  std::vector<float> clip_window{0, 0, 1, 1};
+  
+  // [num_batches, per_batch_size, 4]
```
should the comment be:
// [num_batches, per_batch_size * 4]

azaks2 commented on tensorflow/core/kernels/non_max_suppression_op.cc:
```diff
@@ -195,6 +222,229 @@ void DoNonMaxSuppressionOp(
+  int per_batch_size = total_size_per_batch;
+
+  //perform non_max_suppression operation for each batch independently
+  for (int batch = 0; batch < num_batches; ++batch) {
```
would it make sense to have an auxiliary function to handle each batch

azaks2 commented on tensorflow/core/kernels/non_max_suppression_op.cc:
```diff
@@ -195,6 +222,229 @@ void DoNonMaxSuppressionOp(
+  //perform non_max_suppression operation for each batch independently
+  for (int batch = 0; batch < num_batches; ++batch) {
+    // dims of per_batch_boxes [num_boxes, q, 4]
+    Tensor per_batch_boxes = inp_boxes.Slice(batch, batch+1);
```
spaces around '+'

azaks2 commented on tensorflow/core/kernels/non_max_suppression_op.cc:
```diff
@@ -195,6 +222,229 @@ void DoNonMaxSuppressionOp(
+    std::priority_queue<ResultCandidate, std::vector<ResultCandidate>, 
+      decltype(rc_cmp)> result_candidate_pq(rc_cmp);
+
+    float * scoresData = per_batch_scores.unaligned_flat<float>().data();
```
float* no space

azaks2 commented on tensorflow/core/kernels/non_max_suppression_op.cc:
```diff
@@ -195,6 +222,229 @@ void DoNonMaxSuppressionOp(
+
+    //Iterate through all classes
+    for (int class_idx = 0; class_idx < num_classes; ++class_idx) {
+      std::vector<float> scores_data;
```
name variables consistently.
scores_data
boxes_data
class_scores_data
class_boxes_data

azaks2 commented on tensorflow/core/kernels/non_max_suppression_op.cc:
```diff
@@ -195,6 +222,229 @@ void DoNonMaxSuppressionOp(
+      std::vector<float> boxes_data_vec;
+
+      for (int box = 0; box < num_boxes; ++box) {
+        //Get the scores per class 
```
space after //

azaks2 commented on tensorflow/core/kernels/non_max_suppression_op.cc:
```diff
@@ -195,6 +222,229 @@ void DoNonMaxSuppressionOp(
+        for(int cid = 0; cid < 4; ++cid){
+          if(q > 1){
+            //Get the boxes per class. boxes_data_vec dims is [num_boxes, 4]
+            boxes_data_vec.push_back(boxesData[box * q * 4 + 
```
(box * q + class_idx) * 4 might be a little easier to follow

azaks2 commented on tensorflow/core/kernels/non_max_suppression_op.cc:
```diff
@@ -195,6 +222,229 @@ void DoNonMaxSuppressionOp(
+      auto cmp = [](const Candidate bs_i, const Candidate bs_j) {
+        return bs_i.score < bs_j.score;
+      };
+      std::priority_queue<Candidate, std::deque<Candidate>, decltype(cmp)>
```
why do you need a priority queue (for  candidate_priority_queue and result_candidate_pq) vector/queue should work - you just need to sort them once.

supriyar replied:
I followed the same technique as used by the existing DoNonMaxSuppressionOp() function. It uses a priority queue based on scores. Do you want me to change these?

supriyar replied:
Updated to use vector and sorted it

azaks2 commented on tensorflow/core/kernels/non_max_suppression_op.cc:
```diff
@@ -195,6 +222,229 @@ void DoNonMaxSuppressionOp(
+      for (int box = 0; box < num_boxes; ++box) {
+        //Get the scores per class 
+        //scores_data dim is [num_boxes].
+        scores_data.push_back(scoresData[box * num_classes + class_idx]);
```
can we avoid the copy. especially since you copy back boxes to a tensor afterword

supriyar replied:
I am not sure how to do this since I need to slice along the column of the matrix. Do you have any suggestions on how it can be done?

Commit:
Addressed review comments from PR

Commit:
Use vector instead of priority_queue in kernel op

pooyadavoodi commented:
@azaks2 Could you please check whether there is anything left to do for this PR.

azaks2 commented on tensorflow/core/kernels/non_max_suppression_op_test.cc:
```diff
@@ -861,4 +861,474 @@ TEST_F(NonMaxSuppressionWithOverlapsOpTest, TestEmptyInput) {
+TEST_F(CombinedNonMaxSuppressionOpTest, TestSelectFromThreeClusters) {
+  MakeOp();
+  AddInputFromArray<float>(
+      TensorShape({1, 6, 1, 4}),
```
might be readable to make test data in 6x4 shape?

supriyar replied:
I have some checks in the kernel code which expects batch size to be >=1 so if i set tensor shape to {6,4} it will fail those checks. It also expects boxes to be 4 dimensional in `ParseAndCheckCombinedNMSBoxSizes`

azaks2 commented on tensorflow/core/kernels/non_max_suppression_op_test.cc:
```diff
@@ -861,4 +861,474 @@ TEST_F(NonMaxSuppressionWithOverlapsOpTest, TestEmptyInput) {
+      {0.1f, 0.9f, 0.75f, 0.8f, 0.6f, 0.3f, 0.95f, 0.1f, 0.5f, 0.5f, 0.3f, 0.1f, 
+       0.1f, 0.9f, 0.75f, 0.8f, 0.6f, 0.3f, 0.95f, 0.1f, 0.5f, 0.5f, 0.3f, 0.1f
+       }); 
+  AddInputFromArray<int>(TensorShape({}), {3});
```
can we set larger max so it is clear the value not used

azaks2 commented on tensorflow/core/kernels/non_max_suppression_op_test.cc:
```diff
@@ -861,4 +861,474 @@ TEST_F(NonMaxSuppressionWithOverlapsOpTest, TestEmptyInput) {
+
+TEST_F(CombinedNonMaxSuppressionOpTest,
+        TestSelectFromTwoBatchesTwoClassesWithScoreThresholdPadded) {
+  MakeOp(true);
```
we probably want a few more test with true

supriyar replied:
Added test with padding set to true for max total size and max_size_per_class

azaks2 commented on tensorflow/core/kernels/non_max_suppression_op_test.cc:
```diff
@@ -861,4 +861,474 @@ TEST_F(NonMaxSuppressionWithOverlapsOpTest, TestEmptyInput) {
+                     .Input(FakeInput(DT_FLOAT))
+                     .Attr("pad_per_class", pad_per_class)
+                     .Finalize(node_def()));
+    TF_EXPECT_OK(InitOp());
```
can we run non batched version to check the results.

supriyar replied:
The input values and outputs I've chosen are same as the non batched version.

Commit:
Addressed last set of review comments. Added test for pad_per_class which does
padding based on max_total_size and max_size_per_class

Commit:
Clang and pylint formatting fixes

Commit:
Added api_def_CombinedNonMaxSuppression.pbtxt based on test requirement in
//tensorflow/core/api_def:api_test

azaks2 commented on tensorflow/core/api_def/base_api/api_def_CombinedNonMaxSuppression.pbtxt:
```diff
@@ -0,0 +1,106 @@
+  in_arg {
+    name: "boxes"
+    description: <<END
+A 4-D float tensor of shape `[batch_size, num_boxes, q, 4]`.
```
describe q

azaks2 reviewed (commented):
Please describe the magic q param

Commit:
Added a comment describing 'q'

supriyar commented:
@azaks2 Done, please review

azaks2 commented:
you should run
$ bazel build tensorflow/tools/api/tests:api_compatibility_test
 $ bazel-bin/tensorflow/tools/api/tests/api_compatibility_test \
          --update_goldens True

alextp commented on tensorflow/python/ops/image_ops_impl.py:
```diff
@@ -3058,3 +3058,74 @@ def crop_and_resize_v2(
+      containing the indices of the selected boxes
+  """
+  with ops.name_scope(name, 'combined_non_max_suppression'):
+    iou_threshold = ops.convert_to_tensor(iou_threshold, name='iou_threshold')
```
Specify the float32 dtype here and below to avoid accidentally creating something of the wrong type (or remove these lines since the op binding should ensure that)

supriyar replied:
Done

supriyar commented:
It ran successfully, but didn't generate any new files. Am I missing something?
`INFO: Elapsed time: 1833.833s, Critical Path: 319.09s
INFO: 4249 processes: 4249 local.
INFO: Build completed successfully, 4403 total actions
`
`root@ae98e93f0b3f:/tensorflow_local/tensorflow# bazel-bin/tensorflow/tools/api/tests/api_compatibility_test --update_goldens True
sssWARNING:tensorflow:From /tensorflow_local/tensorflow/bazel-bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/python/util/decorator_utils.py:127: GraphKeys.VARIABLES (from tensorflow.python.framework.ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use `tf.GraphKeys.GLOBAL_VARIABLES` instead.
....
----------------------------------------------------------------------
Ran 7 tests in 0.505s

OK (skipped=3)`

Commit:
Add float dtype to the python API

alextp commented:
It shouldn't generate new files, it should modify the existing ones.

azaks2 commented:
Since we are returning tensors do we really need the selected_indices. we do lose information but it might only be useful if there are some other data that needs to be joined

Commit:
Change golden APIs

pooyadavoodi commented:
I added the changes for golden API.

pooyadavoodi commented:
@azaks2 We don't currently need to return the indices, we added it only for some possible future use cases. If you think it's not necessary to have it now, we can remove it. Please let me know.

azaks2 commented:
> If you think it's not necessary to have it now, we can remove it. Please let me know.
sure lets remove them

Commit:
Removed selected_indices return value for NMS op

pooyadavoodi commented:
@azaks2 The returned indices are removed now. 
Thanks @supriyar

Commit:
Update comment for pad_per_class

supriyar commented on tensorflow/python/ops/image_ops_impl.py:
```diff
@@ -3097,11 +3097,11 @@ def combined_non_max_suppression(boxes,
-      (smaller of the two). Defaults to false.
+    pad_per_class: If false, the output nmsed boxes, scores and classes
+      are padded/clipped to `max_total_size`. If true, the
+      output nmsed boxes, scores and classes are padded/clipped to be of length
```
@azaks2 Updated with the new description.

shahzadlone commented on tensorflow/core/kernels/non_max_suppression_op.cc:
```diff
@@ -195,6 +223,214 @@ void DoNonMaxSuppressionOp(
+      std::vector<float> class_scores_data;
+      std::vector<float> class_boxes_data;
+
+      for (int box = 0; box < num_boxes; ++box) {
```
Before going in this for-loop make sure to 
1) ```class_scores_data.reserve(num_boxes);```
2) ```class_boxes_data.reserve(num_boxes*4);```

Reasoning:
1)Because we know the size, so prefer to reserve in-order to save on reallocation costs.
2)Similar to (1) but note, the way the code flows: it will always push ```num_boxes * 4``` elements into the vector ```class_boxes_data```.

pooyadavoodi replied:
Thanks for the comment. I just pushed a commit that exactly does this.

shahzadlone replied:
Perfect :)

Commit:
Use std::vector::reserve to optimize allocation

azaks2 commented:
Please regenerate api defs for V1 as well - thanks

Commit:
Change golden APIs for V1

pooyadavoodi commented:
@azaks2 Updated the V1 API too. Did it manually because `api_compatibility_test` didn't update it for me for V1.

azaks2 commented on tensorflow/tools/api/golden/v1/tensorflow.pbtxt:
```diff
@@ -932,6 +932,10 @@ tf_module {
     name: "clip_by_value"
     argspec: "args=[\'t\', \'clip_value_min\', \'clip_value_max\', \'name\'], varargs=None, keywords=None, defaults=[\'None\'], "
   }
+  member_method {
```
move it before complex. The list needs to be sorted.

pooyadavoodi replied:
Done. 

Do you know why the compatibility test doesn't work for V1?

azaks2 reviewed (changes requested):
will re approve as soon as sorting is fixed.

Commit:
Move added NMS op in V1 golden API to keep the list sorted

Commit:
Do not set 5th output in CombinedNMSShapeFn

We already removed the 5th output from the op definition. This part
was missed.

aaroey commented:
Hi @pooyadavoodi, would you help to resolve the conflicts? I'll retry the merge after that.
Thanks.

pooyadavoodi commented:
@aaroey I rebased onto master.

shahzadlone commented on tensorflow/core/kernels/non_max_suppression_op.cc:
```diff
@@ -195,6 +223,216 @@ void DoNonMaxSuppressionOp(
+    while (curr_total_size > 0 && result_idx < result_candidate_vec.size()) {
+      ResultCandidate next_candidate = result_candidate_vec[result_idx++];
+      // Add to final output vectors
+      nmsed_boxes[batch].push_back(
```
Maybe reserve this if haven't before: 
1) nmsed_boxes[batch] to be of size (4 * (predicted #iterations of the loop )).
2) nmsed_scores[batch] to be of size (predicted #iterations of the loop ).
3) nmsed_classes[batch] to be of size (predicted #iterations of the loop ).

Where size of iteration will be the minimum between min(curr_total_size, result_candidate_vec.size())
